ciphers/playfair_cipher.c

```diff
@@ -0,0 +1,198 @@
+  int lastUniqueIndex = 0;
+
+  while (*(str + currentIndex)) {
+    char temp = *(str + currentIndex);
+    if (0 == hash[temp]) {
+      hash[temp] = 1;
+      *(str + lastUniqueIndex) = temp;
+      lastUniqueIndex++;
+    }
+    currentIndex++;
+  }
+  *(str + lastUniqueIndex) = '\0';
+}
+
+void playfaircipher(char *str,char *keystr)
+{
+    int i, j, k = 0, l, m = 0, n;
+    char key[MX][MX], keyminus[25];
+    n = strlen(keystr);
+    char alpa[26] = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
+                     'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
+                     'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
+    printf("\t\t\t\tPlayFair Cipher");
+    // convert the characters to uppertext
+    for (i = 0; i < n; i++)
+    {
+        if (keystr[i] == 'j')
+        {
+            keystr[i] = 'i';
+        }
+        else if (keystr[i] == 'J')
+        {
+            keystr[i] = 'I';
+        }
+        keystr[i] = toupper(keystr[i]);
+    }
+
+    // convert all the characters of plaintext to uppertext
+    for (i = 0; i < strlen(str); i++)
+    {
+        if (str[i] == 'j')
+        {
+            str[i] = 'i';
+        }
+        else if (str[i] == 'J')
+        {
+            str[i] = 'I';
+        }
+        str[i] = toupper(str[i]);
+    }
+    // store all characters except key
+    j = 0;
+    for (i = 0; i < 26; i++)
+    {
+        for (k = 0; k < n; k++)
+        {
+            if (keystr[k] == alpa[i])
+            {
+                break;
+            }
+            else if (alpa[i] == 'J')
+            {
+                break;
+            }
+        }
+        if (k == n)
+        {
+            keyminus[j] = alpa[i];
+            j++;
+        }
+    }
+
+    // construct key keymatrix
+    k = 0;
+    for (i = 0; i < MX; i++)
+    {
+        for (j = 0; j < MX; j++)
+        {
+            if (k < n)
+            {
+                key[i][j] = keystr[k];
+                k++;
+            }
+            else
+            {
+                key[i][j] = keyminus[m];
+                m++;
+            }
+            printf("%c ", key[i][j]);
+        }
+        printf("\n");
+    }
+
+    // construct diagram and convert to cipher text
+    printf("\nEntered text :%s\nOutput Text :", str);
+    for (i = 0; i < strlen(str); i++)
+    {
+        if (str[i] == 'J')
+        {
+            str[i] = 'I';
+        }
+        if (str[i + 1] == '\0')
+        {
+            playfair(str[i], 'X', key);
+        }
+        else
+        {
+            if (str[i + 1] == 'J')
+            {
+                str[i + 1] = 'I';
+            }
+            if (str[i] == str[i + 1])
+            {
+                playfair(str[i], 'X', key);
+            }
+            else
+            {
+                playfair(str[i], str[i + 1], key);
+                i++;
+            }
+        }
+    }
+}
+
+void test() 
+{ 
+    char msg[] = "Playfair", key[] = "Good";
+    choice = 1;
+    printf("First Test Case");
```

Comment:
Please refer to the suggestion made for the teaser cipher code on implementing a self-test funcition.